src/nvim/ex_cmds2.c

```diff
@@ -879,10 +882,60 @@ void profile_dump(void)
   }
 }
 
-/*
- * Start profiling script "fp".
- */
-static void script_do_profile(scriptitem_T *si)
+/// Reset all profiling information.
+static void profile_reset(void)
+{
+  // Reset sourced files.
+  for (int id = 1; id <= script_items.ga_len; id++) {
+    scriptitem_T *si = &SCRIPT_ITEM(id);
+    if (si->sn_prof_on) {
+      si->sn_prof_on      = 0;
+      si->sn_pr_force     = 0;
+      si->sn_pr_child     = (proftime_T)0;
+      si->sn_pr_nest      = 0;
+      si->sn_pr_count     = 0;
+      si->sn_pr_total     = (proftime_T)0;
+      si->sn_pr_self      = (proftime_T)0;
+      si->sn_pr_start     = (proftime_T)0;
+      si->sn_pr_children  = (proftime_T)0;
+      ga_clear(&si->sn_prl_ga);
+      si->sn_prl_start    = (proftime_T)0;
+      si->sn_prl_children = (proftime_T)0;
+      si->sn_prl_wait     = (proftime_T)0;
+      si->sn_prl_idx      = -1;
+      si->sn_prl_execed   = 0;
+    }
+  }
+
+  // Reset functions.
+  size_t      n  = func_hashtab.ht_used;
+  hashitem_T *hi = func_hashtab.ht_array;
+
+  for (; n > (size_t)0; hi++) {
+    if (!HASHITEM_EMPTY(hi)) {
+      n--;
+      ufunc_T *uf = HI2UF(hi);
+      if (uf->uf_profiling) {
+        uf->uf_profiling    = 0;
+        uf->uf_tm_count     = 0;
+        uf->uf_tm_total     = (proftime_T)0;
+        uf->uf_tm_self      = (proftime_T)0;
+        uf->uf_tm_children  = (proftime_T)0;
+        uf->uf_tml_count    = NULL;
+        uf->uf_tml_total    = NULL;
+        uf->uf_tml_self     = NULL;
+        uf->uf_tml_start    = (proftime_T)0;
+        uf->uf_tml_children = (proftime_T)0;
+        uf->uf_tml_wait     = (proftime_T)0;
+        uf->uf_tml_idx      = -1;
+        uf->uf_tml_execed   = 0;
+      }
+    }
+  }
```

Comment:
Should also free `profile_fname` and set it to `NULL`. The string is never used again.